3.0/m_relaymsg.cpp

```diff
@@ -0,0 +1,218 @@
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/// $ModAuthor: jlu5
+/// $ModAuthorMail: james@overdrivenetworks.com
+/// $ModDepends: core 3
+/// $ModDesc: Provides the RELAYMSG command & draft/relaymsg capability for stateless bridging
+/// $ModConfig: <relaymsg separators="/" ident="relay" host="relay.example.com">
+//  The "host" option defaults to the local server hostname if not set.
+
+#include "inspircd.h"
+#include "modules/cap.h"
+#include "modules/ircv3.h"
+
+enum
+{
+    ERR_BADRELAYNICK = 573  // from ERR_CANNOTSENDRP in Oragono
+};
+
+// Registers the draft/relaymsg
+class RelayMsgCap : public Cap::Capability {
+public:
+    std::string nick_separators;
+
+    const std::string* GetValue(LocalUser* user) const CXX11_OVERRIDE
+    {
+        return &nick_separators;
+    }
+
+    RelayMsgCap(Module* mod)
+        : Cap::Capability(mod, "draft/relaymsg")
+    {
+    }
+};
+
+// Handler for the @relaymsg message tag sent with forwarded PRIVMSGs
+class RelayMsgCapTag : public ClientProtocol::MessageTagProvider {
+private:
+    RelayMsgCap& cap;
+
+public:
+    RelayMsgCapTag(Module* mod, RelayMsgCap& Cap)
+        : ClientProtocol::MessageTagProvider(mod)
+        , cap(Cap)
+    {
+    }
+
+    bool ShouldSendTag(LocalUser* user, const ClientProtocol::MessageTagData& tagdata) CXX11_OVERRIDE
+    {
+        return cap.get(user);
+    }
+};
+
+// Handler for the RELAYMSG command (users and servers)
+class CommandRelayMsg : public Command {
+private:
+    RelayMsgCap& cap;
+    RelayMsgCapTag& captag;
+    std::bitset<UCHAR_MAX> invalid_chars_map;
+    std::string invalid_chars = "!+%@&#$:'\"?*,.";
+
+public:
+    std::string fake_host;
+    std::string fake_ident;
+
+    CommandRelayMsg(Module* parent, RelayMsgCap& Cap, RelayMsgCapTag& Captag)
+        : Command(parent, "RELAYMSG", 3, 3)
+        , cap(Cap)
+        , captag(Captag)
+    {
+        flags_needed = 'o';
+        syntax = "<channel> <nick> <text>";
+        allow_empty_last_param = false;
+
+        // Populate a map of disallowed nick characters. Based off m_chghost code
+        for (std::string::const_iterator n = invalid_chars.begin(); n != invalid_chars.end(); n++) {
+            invalid_chars_map.set(static_cast<unsigned char>(*n));
+        }
+    }
+
+    std::string GetFakeHostmask(std::string& nick) {
+        return InspIRCd::Format("%s!%s@%s", nick.c_str(), fake_ident.c_str(), fake_host.c_str());
+    }
+
+    CmdResult Handle(User* user, const CommandBase::Params& parameters)
+    {
+        std::string channame = parameters[0];
+        std::string nick = parameters[1];
+        std::string text = parameters[2];
```

Comment:
String copies are unnecessary here.

```suggestion
        const std::string& channame = parameters[0];
        const std::string& nick = parameters[1];
        const std::string& text = parameters[2];
```